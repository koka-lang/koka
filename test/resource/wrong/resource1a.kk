// --------------------------------------------------------
// resources
// test with safe resources using rank-2 types to prevent
// escaping the correct scope
// --------------------------------------------------------

effect heap<s::S> { }

fun heap( action : forall<s> () -> <heap<s>|e> a ) : e a {
  handle<heap<_>>(action) { }
}

effect resource ref<s::S,a::V> in heap<s> {
  fun get() : a
  fun set( value : a ) : ()
}

val new-ref = handler resource(s) {
  get() -> resume(s,s)
  set(x) -> resume((),x)
}

fun foo() {
  use r = new-ref(36)
  r  // ouch, the resource escapes the handler scope
}

fun main() {
  using heap
  val r = foo()
  r.get.println   // this should throw
}
