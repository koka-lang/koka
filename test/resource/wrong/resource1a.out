[01;30moperator branch (new-ref) get: resume tail
[00;39;49m[00m[01;30moperator branch (new-ref) set: resume tail
[00;39;49m[00mfailure while running program:
 node out/test/algeff_wrong_resource1a.js
.../out/test/std_core.js:
 throw result.exn; // rethrow
 ^
Error: there is no handler found for operation ref-resource1a3/get